internal/memory/pool.go

```diff
@@ -0,0 +1,324 @@
+	parent *region
+	class  classType
+	offset uint64
+}
+
+// memoryPool tracks free and busy (used) memory regions
+type memoryPool struct {
+	free map[uint64]*region
+	busy map[uint64]*region
+}
+
+// PoolAllocator implements a memory allocation strategy similar to buddy-malloc https://github.com/evanw/buddy-malloc/blob/master/buddy-malloc.c
+// We borrow the idea of spanning a tree of fixed size regions on top of a contiguous memory
+// space.
+//
+// There are a total of 7 different region sizes that can be allocated, with the smallest
+// being 1MB and the largest 4GB (the default maximum size of a Virtual PMem device).
+//
+// For efficiency and to reduce fragmentation an entire region is allocated when requested.
+// When there's no available region of requested size, we try to allocate more memory for
+// this particular size by splitting the next available larger region into smaller ones, e.g.
+// if there's no region available for size class 0, we try splitting a region from class 1,
+// then class 2 etc, until we are able to do so or hit the upper limit.
+type PoolAllocator struct {
+	pools [memoryClassNumber]*memoryPool
+}
+
+var _ MappedRegion = &region{}
+var _ Allocator = &PoolAllocator{}
+
+func (r *region) Offset() uint64 {
+	return r.offset
+}
+
+func (r *region) Size() uint64 {
+	sz, _ := GetMemoryClassSize(r.class)
+	return sz
+}
+
+func newEmptyMemoryPool() *memoryPool {
+	return &memoryPool{
+		free: make(map[uint64]*region),
+		busy: make(map[uint64]*region),
+	}
+}
+
+func NewPoolMemoryAllocator() PoolAllocator {
+	pa := PoolAllocator{}
+	p := newEmptyMemoryPool()
+	// by default we allocate a single region with maximum possible size (class type)
+	p.free[0] = &region{
+		class:  memoryClassNumber - 1,
+		offset: 0,
+	}
+	pa.pools[memoryClassNumber-1] = p
+	return pa
+}
+
+// FindFreeOffset finds free memory region with minimum offset that can fit a given memory class.
+func (pa *PoolAllocator) FindFreeOffset(memCls classType) (uint64, error) {
+	if memCls >= memoryClassNumber {
+		return 0, ErrInvalidMemoryClass
+	}
+
+	for i := memCls; i < memoryClassNumber; i++ {
+		pi := pa.pools[i]
+		if pi == nil || len(pi.free) == 0 {
+			continue
+		}
+
+		// start off with max offset
+		target := 4 * GigaByte
+		for offset := range pi.free {
+			if offset < target {
+				target = offset
+			}
+		}
+		return target, nil
+	}
+	return 0, ErrNotEnoughSpace
+}
+
+// Allocate checks memory pool of class `memCls` and returns a free region with minimal offset, if none available,
+// tries to expanding current `memCls` pool.
+//
+// Internally it's done via moving a region from free pool into a busy pool
+func (pa *PoolAllocator) Allocate(memCls classType) (MappedRegion, error) {
+	if memCls >= memoryClassNumber {
+		return nil, ErrInvalidMemoryClass
+	}
+
+	// helper function to allocate a region
+	alloc := func() *region {
+		offset, err := pa.FindFreeOffset(memCls)
+		if err != nil {
+			return nil
```

Comment:
I think this should be return err

Reply:
Oh I see, sorry. I think this this should probably return an error too. We can choose to ignore whatever error, but I think it should at least be propagated to the caller.

Reply:
did some refactoring around it, ptal 😄